### ISSUE TITLE
spec: constraint type inference should proceed even if an individual unification step fails

### ISSUE THREAD

**Issue by RelicOfTesla** [golang/go#53650]
Edit: See https://github.com/golang/go/issues/53650#issuecomment-1334495063 for a summary of the problem.

### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.18.3 windows/amd64
</pre>

### What did you do?

```go

type BaseT interface {
	Type1 | Type2
}
type BaseType int
type Type1 BaseType
type Type2 BaseType // float64

type ValueT[T BaseT] struct {
	A1 T
}

func NewType1() *ValueT[Type1] {
	r := NewT[Type1]()
	return r
}
func NewType2() *ValueT[Type2] {
	r := NewT[Type2]()
	return r
}

func NewT[TBase BaseT, TVal ValueT[TBase]]() *TVal {
	ret := TVal{}
	return &ret
}
func TestGoType(t *testing.T) {
	r1 := NewType1()
	r2 := NewType2()
	t.Log(r1, r2)
	t.Log(reflect.TypeOf(r1), reflect.TypeOf(r2))
	fooT1(r1.A1)
	fooT2(r2.A1)
}

func fooT1(t1 Type1) {

}
func fooT2(t2 Type2) {

}

```

### What did you expect to see?

Compile success

### What did you see instead?

error:  
```
TBase does not match int
TBase does not match int
```

**Comment by ianlancetaylor** [comment]
CC @griesemer 

Standalone reproducer:

```Go
package p

type BaseT interface {
	Type1 | Type2
}
type BaseType int
type Type1 BaseType
type Type2 BaseType

type ValueT[T BaseT] struct {
	A1 T
}

func NewType1() *ValueT[Type1] {
	r := NewT[Type1]()
	return r
}

func NewT[TInt BaseT, TVal ValueT[TInt]]() *TVal {
	ret := TVal{}
	return &ret
}
```

```
foo.go:15:18: TInt does not match int
```

**Comment by hopehook** [comment]
I'm not good at generics, maybe there is a flaw here. And I find the compiler will be happy with these cases:

Case one based on @RelicOfTesla's code:

```go
type BaseT interface {
	Type1 | Type2
}
type BaseType int
type Type1 BaseType
type Type2 BaseType // float64

type ValueT[T BaseT] struct {
	A1 T
}

func NewType1() *ValueT[Type1] {
	r := NewT[Type1, ValueT[Type1]]()
	return r
}
func NewType2() *ValueT[Type2] {
	r := NewT[Type2, ValueT[Type2]]()  // add `, ValueT[Type2]`
	return r
}

func NewT[TBase BaseT, TVal ValueT[TBase]]() *TVal {
	ret := TVal{}
	return &ret
}
func TestGoType(t *testing.T) {
	r1 := NewType1()
	r2 := NewType2()
	t.Log(r1, r2)
	t.Log(reflect.TypeOf(r1), reflect.TypeOf(r2))
	fooT1(r1.A1)
	fooT2(r2.A1)
}

func fooT1(t1 Type1) {

}
func fooT2(t2 Type2) {

}
```


Case two based on @ianlancetaylor's code:

```go
package p

type BaseT interface {
	Type1 | Type2
}
type BaseType int
type Type1 BaseType
type Type2 BaseType

type ValueT[T BaseT] struct {
	A1 T
}

func NewType1() *ValueT[Type1] {
	r := NewT[Type1, ValueT[Type1]]()  // add `, ValueT[Type1]`
	return r
}

func NewT[TInt BaseT, TVal ValueT[TInt]]() *TVal {
	ret := TVal{}
	return &ret
}
```

Case three is modified by me:

```
package p

type BaseT interface {
	Type1 | Type2
}

type Type1 int
type Type2 int

var _ = NewT[Type1, Type1]()  // add `Type1` or `Type2`

func NewT[TInt, TVal BaseT]() TVal {
	return TVal(0)
}
```

From the results, it seems that we do not support indirect type deduction `TVal` now.

**Comment by griesemer** [comment]
Smaller reproducer:
```Go
type T1 int
type T2 int // code works if this is changed to, say int8

type G[_ T1 | T2] struct{}

func _() {
	f[T1]()
	f[T2]()
}

func f[P T1 | T2, Q G[P]]() {}
```

**Comment by griesemer** [comment]
Even smaller reproducer:
```Go
type T1 int
type T2 int

func f[P T1 | T2, _ []P]() {}

var _ = f[T1]
```
This is a bug with constraint type inference: `P` is assigned the type `T1`. Constraint type inference kicks in to determine the type of the 2nd type argument (it doesn't matter what it is). In order to proceed, constraint type inference attempts to unify the type argument `T1` for `P` with the (adjusted) core type of `P`'s constraint `T1 | T2`. That core type is `int`. But `int` and `T1` cannot be unified (they are different, named types) and we get an (incorrect) error.

If `T1` and `T2` are defined as (for instance) `struct{}`, the core type is `struct{}` and `T1` can be unified with `struct{}`.

If `T1` and `T2` have different underlying types, no core type exists, and no unification is attempted with the first type parameter and its constraint. Inference proceeds with the 2nd type parameter and the result is `[]T1` in that case.

So constraint type inference is a bit too aggressive here. This area is notoriously tricky. May not be safe to address at this point in the release cycle (TBD). Thinking.

For the release team: It's safe to kick this down to Go 1.21 if we don't find a safe fix in time. This is not a regression; the bug existed since 1.18. Also, providing all type arguments manually is a safe work-around.

FYI @findleyr

**Comment by griesemer** [comment]
Per discussion with @findleyr, we believe https://go.dev/cl/454655 is almost certainly correct, but there's no rush to fix this for 1.20. Still thinking about possible ramifications.

**Comment by griesemer** [comment]
In the [section on constraint type inference](https://tip.golang.org/ref/spec#Constraint_type_inference), the spec states:

> 1. For all type parameters with an adjusted core type, unify the type parameter with that type. **If any unification fails, constraint type inference fails.**

which is exactly what the implementation is doing. The fix (https://go.dev/cl/454655) does something else:

> 1. For all type parameters with an adjusted core type, unify the type parameter with that type. **Proceed even if the unification fails.**

I believe this new approach is the correct approach (and does fix this issue), but it is in fact a (backward-compatible, I think) spec change. In other words, the current behavior is "correct" per the spec.

Making this a spec change and moving it to 1.21.
